src/main/java/com/researchspace/service/impl/RecordManagerImpl.java

```diff
@@ -1198,20 +1182,29 @@ private List<BaseRecord> getViewableRecordsByRole(User subject, boolean onlyTemp
        *
        */
       List<User> viewableUsers = userDao.getViewableUsersByRole(subject);
-      Set<BaseRecord> recordsToAdd = new HashSet<>();
+
+      // records shared with the subject
+      Set<BaseRecord> subjectsViewableRecords =
+          new HashSet<>(getRecordsSharedWithUser(subject, onlyTemplates, subject));
+
+      // records created by each of the subjects viewable users
       for (User viewableUser : viewableUsers) {
-        recordsToAdd.addAll(getRecordsSharedWithUser(viewableUser, onlyTemplates, subject));
-        recordsToAdd.addAll(getViewableRecordsByUser(viewableUser, onlyTemplates));
+        subjectsViewableRecords.addAll(
+            getViewableRecordsByUsers(List.of(viewableUser), onlyTemplates));
+        if (viewableUser.hasRole(Role.PI_ROLE)) {
+          // if the subject is a PI of group A, which contains another PI as a group member, they
+          // can only view the docs of the other PI which has been shared with group A (whereas the
+          // subject PI can view ALL records created by a non-PI), so filter the other PIs records
+          // by their permissions to ensure correct visibility
+          subjectsViewableRecords =
+              subjectsViewableRecords.stream()
+                  .filter(br -> permissnUtils.isPermitted(br, PermissionType.READ, subject))
+                  .collect(toCollection(HashSet::new));
```

Comment:
Here qwe are re-createing the set and reassign every loop.
is there any chance we could just remove from the original set the ones that are not permitted?

Reply:
Good point. The reason for this probably inefficient looping was to cover the case where the PI of
group A can see the records of their group member, who is a PI of group B. In that case, PI A should only be able to see the records shared with group A by PI B, and none of their other work in group B.

But by doing so we are individually fetching each users records from the db and not making use of the new method which takes the set of user ids. I'll add instead just a single loop at the end through all records, and if the owner of that record is a PI, check permissions and remove if not permitted. Should speed things up a bit, thanks!